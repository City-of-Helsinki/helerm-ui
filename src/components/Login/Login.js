import React, { PropTypes } from 'react';

class Login extends React.Component {
  constructor () {
    super();

    this.handleUserLinkClick = this.handleUserLinkClick.bind(this);
  }

  handleUserLinkClick (e) {
    e.preventDefault();
    const { user } = this.props;
    const linkMethod = user.id ? this.props.logout : this.props.login;
    linkMethod();
  }

  getUserLink () {
    const { user } = this.props;
    const linkText = (user && user.id) ? 'Kirjaudu ulos' : 'Kirjaudu sisään';

    return (
      <a href='' className='navbar-link' onClick={this.handleUserLinkClick}>{linkText}</a>
    );
  }

  render () {
    const { user } = this.props;
    const displayName = (user && user.id) ? user.displayName ? user.displayName : `${user.firstName}${user.lastName ? ' ' + user.lastName : ''}` : null;
    return (
      <p className='navbar-text pull-right login-link'>
        {!!displayName && <small>{displayName}</small>}
        {this.getUserLink()}
      </p>
    );
  }
}

Login.propTypes = {
  login: PropTypes.func.isRequired,
  logout: PropTypes.func.isRequired,
  user: PropTypes.object
};

export default Login;
