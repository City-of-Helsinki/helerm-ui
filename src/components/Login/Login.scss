.login-container {
  float: right !important;
  margin-right: 0 !important;
}
