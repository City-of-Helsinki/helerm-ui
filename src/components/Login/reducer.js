/* eslint-disable import/no-named-as-default-member */
import update from 'immutability-helper';
import { createAction, handleActions } from 'redux-actions';
import { get } from 'lodash';
import { getApiTokenFromStorage } from 'hds-react'

import api, { Unauthorized } from '../../utils/api';
import { USER_LOGIN_STATUS } from '../../constants'
import config from '../../config';

const initialState = {
  data: null,
  isFetching: false,
  status: USER_LOGIN_STATUS.NONE,
};

export const RECEIVE_USERDATA = 'receiveUserDataAction';
export const RETRIEVE_USERDATA = 'retrieveUserFromSessionAction';
export const ERROR_USERDATA = 'errorUserDataAction';
export const LOGIN_STATUS = 'loginStatusAction';

export function receiveUserData(user) {
  return createAction(RECEIVE_USERDATA)(user);
}

export function retrieveUserFromSession(id) {
  return (dispatch) => {
    const token = getApiTokenFromStorage(config.API_TOKEN_AUTH_AUDIENCE);

    if (!id || !token) {
      dispatch(createAction(ERROR_USERDATA)());
      dispatch(createAction(LOGIN_STATUS)(USER_LOGIN_STATUS.NONE));
      return null
    }
    dispatch(createAction(RETRIEVE_USERDATA)());
    dispatch(createAction(LOGIN_STATUS)(USER_LOGIN_STATUS.INITIALIZING));
    const url = `user/${id}`;
    return api
      .get(url)
      .then((helermUserData) => {
        if (helermUserData.status === 401) {
          throw new Unauthorized(url)
        }
        return helermUserData.json();
      })
      .then((helermUser) => {
        const permissions = get(helermUser, 'permissions', []);
        const userWithPermissions = {
          id,
          firstName: helermUser.first_name,
          lastName: helermUser.last_name,
          permissions,
        };
        return dispatch(receiveUserData(userWithPermissions));
      })
      .catch(() => {
        dispatch(createAction(ERROR_USERDATA)());
      });
  };
}

const retrieveUserFromSessionAction = (state) =>
  update(state, {
    isFetching: { $set: true }
  });

const receiveUserDataAction = (state, { payload }) => update(state, {
  data: { $set: payload },
  isFetching: { $set: false }
});

export default handleActions(
  {
    retrieveUserFromSessionAction,
    receiveUserDataAction
  },
  initialState
);
