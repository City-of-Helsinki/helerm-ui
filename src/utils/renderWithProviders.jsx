/* eslint-disable react/prop-types */
import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { LoginProvider } from 'hds-react';

import { initialState as navigationInitialState } from '../store/reducers/navigation';
import { initialState as validationInitialState } from '../store/reducers/validation';
import { initialState as routerInitialState } from '../store/reducers/router';
import { initialState as selectedTOSInitialState } from '../store/reducers/tos-toolkit/main';
import { initialState as classificationInitialState } from '../store/reducers/classification';
import { initialState as userInitialState } from '../store/reducers/user';
import { initialState as uiInitialState } from '../store/reducers/ui';
import { initialState as bulkInitialState } from '../store/reducers/bulk';
import { initialState as searchInitialState } from '../store/reducers/search';
import storeCreator from '../store/createStore';

export const storeDefaultState = {
  navigation: navigationInitialState,
  validation: validationInitialState,
  router: routerInitialState,
  selectedTOS: selectedTOSInitialState,
  classification: classificationInitialState,
  user: userInitialState,
  ui: uiInitialState,
  bulk: bulkInitialState,
  search: searchInitialState,
};

const renderWithProviders = (
  ui,
  { preloadedState = storeDefaultState, store = storeCreator(preloadedState), ...renderOptions } = {},
) => {
  const Wrapper = ({ children }) => (
    <LoginProvider>
      <Provider store={store}>{children}</Provider>
    </LoginProvider>
  );

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};

export default renderWithProviders;
