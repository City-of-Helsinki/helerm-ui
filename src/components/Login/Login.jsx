/* eslint-disable camelcase */
import React, { useEffect, useState } from 'react';
import './Login.scss';
import { useOidcClient } from 'hds-react';

const Login = () => {
  const [displayName, setDisplayName] = useState(null);

  const { isAuthenticated, login, logout, getUser } = useOidcClient();

  const user = getUser();
  const isAuth = isAuthenticated();

  const getDisplayName = (profile) => {
    const { name, family_name, given_name } = profile;

    if (name) {
      return name;
    }

    if (given_name) {
      if (family_name) {
        return `${given_name} ${family_name}`;
      }

      return given_name;
    }

    return null;
  };

  useEffect(() => {
    if (isAuth) {
      const { profile } = user;

      const userName = getDisplayName(profile);

      setDisplayName(userName);
    }
  }, [user, isAuth]);

  const handleUserLinkClick = (event) => {
    event.preventDefault();

    if (isAuth) {
      logout();
    } else {
      login();
    }
  };

  return (
    <p className='navbar-text pull-right login-link'>
      {!!displayName && <small>{displayName}</small>}

      <button className='btn btn-link login-button' type='button' onClick={handleUserLinkClick}>
        {isAuth ? 'Kirjaudu ulos' : 'Kirjaudu sisään'}
      </button>
    </p>
  );
};

export default Login;
