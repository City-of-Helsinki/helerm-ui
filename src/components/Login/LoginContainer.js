import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import {
  login,
  logout,
  retrieveUserFromSession
} from './reducer';

import Login from './Login';

const mapDispatchToProps = (dispatch) => {
  return {
    login: bindActionCreators(login, dispatch),
    logout: bindActionCreators(logout, dispatch),
    retrieveUserFromSession: bindActionCreators(retrieveUserFromSession, dispatch)
  };
};

const mapStateToProps = (state) => {
  return {
    user: state.user
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Login);
