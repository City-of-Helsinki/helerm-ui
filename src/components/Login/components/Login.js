import React, { PropTypes } from 'react';

import '../Login.scss';

class Login extends React.Component {
  constructor () {
    super();

    this.handleUserLinkClick = this.handleUserLinkClick.bind(this);
  }

  componentWillMount () {
    this.props.retrieveUserFromSession();
  }

  handleUserLinkClick (e) {
    e.preventDefault();
    const { user } = this.props;
    const linkMethod = user.id ? this.props.logout : this.props.login;
    linkMethod();
  }

  getUserLink () {
    const { user } = this.props;
    const linkText = user.id ? `${user.firstName}, Kirjaudu ulos` : 'Kirjaudu sisään';

    return (
      <a href='' className='navbar-link' onClick={this.handleUserLinkClick}>{linkText}</a>
    );
  }

  render () {
    return (
      <p className='navbar-text login-container'>
        {this.getUserLink()}
      </p>
    );
  }
}

Login.propTypes = {
  login: PropTypes.func.isRequired,
  logout: PropTypes.func.isRequired,
  retrieveUserFromSession: PropTypes.func.isRequired,
  user: PropTypes.object.isRequired
};

export default Login;
