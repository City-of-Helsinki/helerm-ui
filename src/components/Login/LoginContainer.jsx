import { bindActionCreators } from '@reduxjs/toolkit';
import { connect } from 'react-redux';

import { login, logout } from './reducer';
import Login from './Login';

const mapDispatchToProps = (dispatch) => ({
  login: bindActionCreators(login, dispatch),
  logout: bindActionCreators(logout, dispatch),
});

const mapStateToProps = (state) => ({
  user: state.user.data,
});

export default connect(mapStateToProps, mapDispatchToProps)(Login);
