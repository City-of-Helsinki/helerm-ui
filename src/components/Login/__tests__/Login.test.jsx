import React from 'react';
import { Router } from 'react-router-dom';
import * as mockLogin from 'hds-react';
import { screen } from '@testing-library/react';
import { createBrowserHistory } from 'history';
import userEvent from '@testing-library/user-event';

import Login from '../Login';
import renderWithProviders from '../../../utils/renderWithProviders';

const getUserMock = vitest.fn().mockImplementation(() => ({ profile: { name: 'Test Tester' } }));
const isAuthenticatedMock = vitest.fn().mockImplementation(() => true);
const loginMock = vitest.fn();
const logoutMock = vitest.fn();

vitest.spyOn(mockLogin, 'useOidcClient').mockImplementation(() => ({
  getUser: getUserMock,
  isAuthenticated: isAuthenticatedMock,
  login: loginMock,
  logout: logoutMock,
}));

const renderComponent = (history) =>
  renderWithProviders(
    <Router history={history}>
      <Login />
    </Router>,
    { history },
  );

describe('<Login />', () => {
  it('should render correctly', () => {
    const history = createBrowserHistory();
    const { container } = renderComponent(history);

    expect(container).toMatchSnapshot();
  });

  it('should display user name with link', async () => {
    const history = createBrowserHistory();
    renderComponent(history);

    expect(await screen.findByText(/Test Tester/)).toBeInTheDocument();
    expect(await screen.findByText(/Kirjaudu ulos/)).toBeInTheDocument();
  });

  it('should display given name with family name', async () => {
    vitest.spyOn(mockLogin, 'useOidcClient').mockImplementationOnce(() => ({
      getUser: vitest
        .fn()
        .mockImplementation(() => ({ profile: { given_name: 'Test', family_name: 'Tester', name: undefined } })),
      isAuthenticated: isAuthenticatedMock,
    }));

    const history = createBrowserHistory();
    renderComponent(history);

    expect(await screen.findByText(/Test Tester/)).toBeInTheDocument();
  });

  it('should display just given name', async () => {
    vitest.spyOn(mockLogin, 'useOidcClient').mockImplementationOnce(() => ({
      getUser: vitest
        .fn()
        .mockImplementation(() => ({ profile: { given_name: 'Test', family_name: undefined, name: undefined } })),
      isAuthenticated: isAuthenticatedMock,
    }));

    const history = createBrowserHistory();
    renderComponent(history);

    expect(await screen.findByText(/Test/)).toBeInTheDocument();
  });

  it('should not display user if not authenticated', async () => {
    vitest.spyOn(mockLogin, 'useOidcClient').mockImplementationOnce(() => ({
      getUser: vitest.fn().mockImplementation(() => ({ profile: { name: 'Test Tester' } })),
      isAuthenticated: vitest.fn().mockImplementation(() => false),
    }));

    const history = createBrowserHistory();
    renderComponent(history);

    expect(screen.queryByText(/Test Tester/)).not.toBeInTheDocument();
    expect(await screen.findByRole('button', { name: /Kirjaudu sis채채n/ })).toBeInTheDocument();
  });

  it('should login', async () => {
    vitest.spyOn(mockLogin, 'useOidcClient').mockImplementationOnce(() => ({
      getUser: getUserMock,
      login: loginMock,
      isAuthenticated: vitest.fn().mockImplementation(() => false),
    }));

    const history = createBrowserHistory();
    renderComponent(history);

    const loginLink = await screen.findByRole('button', { name: /Kirjaudu sis채채n/ });
    const user = userEvent.setup();

    await user.click(loginLink);

    expect(loginMock).toHaveBeenCalled();
  });

  it('should login', async () => {
    vitest.spyOn(mockLogin, 'useOidcClient').mockImplementationOnce(() => ({
      getUser: getUserMock,
      logout: logoutMock,
      isAuthenticated: isAuthenticatedMock,
    }));

    const history = createBrowserHistory();
    renderComponent(history);

    const logoutLink = await screen.findByRole('button', { name: /Kirjaudu ulos/ });
    const user = userEvent.setup();

    await user.click(logoutLink);

    expect(logoutMock).toHaveBeenCalled();
  });
});
