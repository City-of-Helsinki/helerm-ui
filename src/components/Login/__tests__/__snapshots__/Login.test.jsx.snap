// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Login /> > should render correctly 1`] = `
<div>
  <p
    class="navbar-text pull-right login-link"
  >
    <small>
      Test Tester
    </small>
    <button
      class="btn btn-link login-button"
      type="button"
    >
      Kirjaudu ulos
    </button>
  </p>
</div>
`;
